r&d.md

```diff
@@ -1276,19 +1304,202 @@
+    - sometimes it takes native object and updates it, ignoring return `:style="style => style.x = 123"`
+      ~+ we can make arg immutable, `obj` for style, `str` for class.
+    - `:fx="a => b => c"` vs `:fx="..., b => c"`
+    - `:ref="el => (mount, () => unmount)"`
+    - `:ona..onb="..., b => c"`
+
+## [x] How to detect function vs direct code `:x="x=>x"` vs `:x=x`? -> let's just try invoking the result if that's a function
+
+  - generally mixing things up like that is generally a mess (sorry vue, that's f*d up)
+
+  0. If effect returns a function, we call that function (continuation of normal effect flow) that's it.
+    * We can create a subeffect for returned function
+    + that makes events both inline or by-method
+    + that enables other effects to be functions
+    ~- that screws up fx cleanup, since it becomes `:fx="() => () => cleanup()"`
+      - `:fx` doesn't have argument, so we enforce fn without a good reason
+
+  1. Make all directives via events
+    + unified code
+    - `:onx` event vs `:x` property becomes the same, not different
+      -~ `:click` and `:onclick` are potentially different things
+      +~ but they never intersect in HTML, besides spread is there for that
+    - possibly a bit heavier - triggering events for props
+    + allows subscribing directly to props change, like `onstyle`, `onclass`, `ontext`, `onif`!, `oneach` - useful!
+      + `.emit` modifier for that purpose
+
+  2. `:on`, `@` for events, `:` for else
+    - `:` props has code that detects a function, same as event
+    + `@` allows binding multiple events to props change
+
+  3. detect statically
+    - doesn't detect property from state `:x="xEffectFn"`
+    + allows unified way to pre-parse expr, as `:each` does
+
+## [x] Should directives return `dispose` (instead of forcing effect `update` loop)? -> no, we need to on/off effects, so we need update fn
+  + generally speaking directives don't have to follow single-effect update schema
+    + eg. events don't need to react to every effect change
+    + scope, fx doesn't have a return
+    + directives can be async, interval etc - they generally define behavior
+    - we decide async logic my modifiers, not arbitrarily by effects. Effects are supposed to be direct sync call.
+  - takes more space
+  - we should have control over disabling / re-enabling effects (on/off), so we need update fn as well
+  - we control evaluation via standard mods, so we need update
+
+## [ ] Should we have a separate event handler like `@`, or that's just a kind of directive?
+
+  + allows easier alpine, vue entries
+  + allows cleaner `sprae` code
+  - performs extra check per-attribute
+
+## [x] Should we register full name of directive `:x`? -> no, let's keep things simple
+  - No way to customize prefix
+    - Complicates JSX entry
+  - No much perf profit if not less
+
+## [x] Flat directives.js instead of directives/ -> yes
```

Comment:
hm, ok, i see that as the explanation to my comment above, but i somewhat disagree, i was for example thinking about an `:edit` directive which i would only register for users with admin roles, e.g. to build something like an "in-page" editor for some JSON config or whatever.

Reply:
You still will be able to add your own directives as `sprae.dir.edit = (el, state, expression) => {...init; return (values) => {...update; return () => ...dispose}}`.
The comment is more about merging directives into a single file.
But it's WIP, likely there won't even be a directives.js, just one sprae.js standard bundle. Because directives are relatively short and very interconnected, splitting to files creates unnecessary code glue that inflates bundle.